src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs

```diff
@@ -90,19 +84,23 @@ private static bool IsValueWriteAtomic()
             //
             Type valueType = typeof(TValue);
             bool isAtomic =
-                   valueType == typeof(bool)
-                || valueType == typeof(char)
-                || valueType == typeof(byte)
-                || valueType == typeof(sbyte)
-                || valueType == typeof(short)
-                || valueType == typeof(ushort)
-                || valueType == typeof(int)
-                || valueType == typeof(uint)
-                || valueType == typeof(float);
+                !valueType.GetTypeInfo().IsValueType ||
+                valueType == typeof(bool) ||
+                valueType == typeof(char) ||
+                valueType == typeof(byte) ||
+                valueType == typeof(sbyte) ||
+                valueType == typeof(short) ||
+                valueType == typeof(ushort) ||
+                valueType == typeof(int) ||
+                valueType == typeof(uint) ||
+                valueType == typeof(float);
 
             if (!isAtomic && IntPtr.Size == 8)
             {
-                isAtomic = valueType == typeof(double) || valueType == typeof(long);
+                isAtomic =
+                    valueType == typeof(double) ||
+                    valueType == typeof(long) ||
+                    valueType == typeof(ulong);
```

Comment:
@stephentoub where did this ulong here come from? I didn't see any notes about its inclusion.

Reply:
sorry for resurrecting your old PR by the way. I'm working on backporting this to netfx :)

Reply:
Just rounding out the list of valid types that meet the criteria.

Reply:
Doesn't it make sense to also include `DateTime` in that list? It's pretty commonly used in dictionaries and seems to be just the right size ([Can DateTime tear in a 64 bit environment?](https://stackoverflow.com/q/42258346/885318))

Reply:
https://github.com/dotnet/corefx/pull/16331#discussion_r102117158